File: ts/src/model-normalization.ts
const KNOWN_PROVIDERS = new Set([
  'anthropic',
  'azure',
  'azure-openai-responses',
  'bedrock',
  'deepseek',
  'gemini',
  'google',
  'groq',
  'mistral',
  'ollama',
  'openai',
  'openai-codex',
  'openrouter',
  'qwen',
  'vertex',
  'xai',
])

const BARE_MODEL_HARNESSES = new Set([
  'claude-code',
  'codex',
  'continue-cli',
  'gemini',
  'openclaude',
  'qwen',
])

const PROVIDER_MODEL_HARNESSES = new Set([
  'aider',
  'kilo',
  'opencode',
  'swe-agent',
])

const PRESERVE_EXPLICIT_PROVIDER_HARNESSES = new Set([
  'crush',
])

export function stripKnownProviderPrefixes(model: string): string {
  let normalized = model.trim()
  while (true) {
    const slash = normalized.indexOf('/')
    if (slash < 0) return normalized
    const head = normalized.slice(0, slash).toLowerCase()
    if (!KNOWN_PROVIDERS.has(head)) return normalized
    normalized = normalized.slice(slash + 1).trim()
  }
}

export function inferProviderForModel(model: string, defaultProvider = 'openai'): string {
  const m = model.trim().toLowerCase()
  if (!m) return defaultProvider
  if (m === 'sonnet' || m === 'opus' || m === 'haiku') return 'anthropic'
  if (m.startsWith('claude') || m.startsWith('sonnet') || m.startsWith('opus') || m.startsWith('haiku')) return 'anthropic'
  if (m.startsWith('gemini')) return 'google'
  if (m.startsWith('qwen')) return 'qwen'
  if (m.startsWith('deepseek')) return 'deepseek'
  if (m.startsWith('grok')) return 'xai'
  if (m.startsWith('mistral')) return 'mistral'
  return defaultProvider
}

export function ensureProviderPrefix(model: string, defaultProvider = 'openai'): string {
  const normalized = model.trim()
  if (!normalized) return normalized

  const slash = normalized.indexOf('/')
  if (slash > 0) {
    const head = normalized.slice(0, slash).toLowerCase()
    if (KNOWN_PROVIDERS.has(head)) return normalized
  }

  const provider = inferProviderForModel(normalized, defaultProvider)
  return `${provider}/${normalized}`
}

export interface NormalizeOptions {
  resolve?: boolean
}

/**
 * Normalize a model identifier for the target harness CLI.
 *
 * This is intentionally best-effort, not an exhaustive provider registry.
 * Callers can bypass normalization entirely by passing `{ resolve: false }`.
 */
export function normalizeModelForHarness(
  harness: string,
  model: string | undefined,
  opts: NormalizeOptions = {},
): string | undefined {
  if (model === undefined) return undefined
  const resolve = opts.resolve ?? true
  const normalized = model.trim()
  if (!normalized || !resolve) return normalized

  if (harness === 'pi') {
    if (normalized.includes('/')) {
      return ensureProviderPrefix(normalized, 'openai-codex')
    }
    if (normalized.toLowerCase().startsWith('gpt-5')) {
      return ensureProviderPrefix(normalized, 'openai-codex')
    }
    return normalized
  }
  if (harness === 'factory-droid') {
    // droid only reaches a non-Factory endpoint via BYOK custom models,
    // which require the "custom:" prefix on the model id. We expect the
    // caller to have a matching entry in ~/.factory/settings.json.
    const bare = stripKnownProviderPrefixes(normalized)
    if (bare.startsWith('custom:')) return bare
    return `custom:${bare}`
  }
  if (PROVIDER_MODEL_HARNESSES.has(harness)) {
    return ensureProviderPrefix(normalized)
  }
  if (PRESERVE_EXPLICIT_PROVIDER_HARNESSES.has(harness) && normalized.includes('/')) {
    return normalized
  }
  if (BARE_MODEL_HARNESSES.has(harness)) {
    return stripKnownProviderPrefixes(normalized)
  }
  return normalized
}

File: src/harness/model_normalization.py
"""Shared model normalization helpers.

Goal: let callers pass canonical names like ``gpt-5.4`` across harnesses while
still emitting each CLI's preferred model format.
"""
from __future__ import annotations

KNOWN_PROVIDERS = {
    "anthropic",
    "azure",
    "bedrock",
    "deepseek",
    "gemini",
    "google",
    "groq",
    "mistral",
    "ollama",
    "openai",
    "openrouter",
    "qwen",
    "vertex",
    "xai",
}

BARE_MODEL_HARNESSES = {
    "claude-code",
    "codex",
    "continue-cli",
    "factory-droid",
    "gemini",
    "openclaude",
    "pi",
    "qwen",
    "crush",
}

PROVIDER_MODEL_HARNESSES = {
    "aider",
    "kilo",
    "opencode",
    "swe-agent",
}


def strip_known_provider_prefixes(model: str) -> str:
    """Remove one or more known provider prefixes from ``model``.

    Examples:
      - ``openai/gpt-5.4`` -> ``gpt-5.4``
      - ``openrouter/anthropic/claude-sonnet-4-6`` -> ``claude-sonnet-4-6``
    """
    normalized = model.strip()
    while True:
        head, sep, tail = normalized.partition("/")
        if not sep:
            return normalized
        if head.lower() not in KNOWN_PROVIDERS:
            return normalized
        normalized = tail.strip()


def infer_provider_for_model(model: str, default_provider: str = "openai") -> str:
    """Best-effort provider inference for canonical (bare) model names."""
    m = model.strip().lower()

    if not m:
        return default_provider
    if m in {"sonnet", "opus", "haiku"}:
        return "anthropic"
    if m.startswith(("claude", "sonnet", "opus", "haiku")):
        return "anthropic"
    if m.startswith("gemini"):
        return "google"
    if m.startswith("qwen"):
        return "qwen"
    if m.startswith("deepseek"):
        return "deepseek"
    if m.startswith("grok"):
        return "xai"
    if m.startswith("mistral"):
        return "mistral"
    return default_provider


def ensure_provider_prefix(model: str, default_provider: str = "openai") -> str:
    """Ensure ``provider/model`` format for CLIs that require it."""
    normalized = model.strip()
    if not normalized:
        return normalized

    head, sep, _tail = normalized.partition("/")
    if sep and head.lower() in KNOWN_PROVIDERS:
        return normalized

    provider = infer_provider_for_model(normalized, default_provider=default_provider)
    return f"{provider}/{normalized}"


def normalize_model_for_harness(harness: str, model: str | None) -> str | None:
    """Normalize model identifier for the target harness CLI."""
    if model is None:
        return None

    normalized = model.strip()
    if not normalized:
        return normalized

    if harness in PROVIDER_MODEL_HARNESSES:
        return ensure_provider_prefix(normalized, default_provider="openai")
    if harness in BARE_MODEL_HARNESSES:
        return strip_known_provider_prefixes(normalized)
    return normalized
